parse_string.go

```diff
@@ -29,59 +20,55 @@ func (p *parser) parseString() (Node, error) {
 		return nil, p.makeError("expecting string delimiter")
 	}
 
-	var str string
-
-	var inEscape bool
-	var inHex bool
-	var hex []rune
-
+	// Start with opening quote
+	buffer := []byte{'"'}
 	for {
 		c, err := p.next()
 		if err != nil {
 			return nil, err
 		}
-		if inHex {
-			switch {
-			case c >= 'a' && c <= 'f', c >= 'A' && c <= 'F', c >= '0' && c <= '9':
-				hex = append(hex, c)
-				if len(hex) == 4 {
-					inHex = false
-					char, err := strconv.Unquote(fmt.Sprintf("'\\u%s'", string(hex)))
-					if err != nil {
-						return nil, p.makeError("invalid unicode character '%s'", err)
-					}
-					str += char
-					hex = nil
-				}
-			default:
-				return nil, p.makeError("invalid hexedecimal escape sequence '\\%s%c'", string(hex), c)
-			}
-		} else if inEscape {
-			inEscape = false
-			if c == 'u' {
-				inHex = true
-				continue
-			}
-			seq, ok := escapes[c]
-			if !ok {
-				return nil, p.makeError("invalid escape sequence '\\%c'", c)
-			}
-			str += seq
-		} else {
-			switch c {
-			case '\\':
-				inEscape = true
-			case '"':
-				n.raw = str
-				n.end = p.position
-				return n, nil
-			default:
-				if c < 0x20 || c > 0x10FFFF {
-					return nil, p.makeError("invalid unescaped character '0x%X'", c)
-				}
-				str += string(c)
+		if c == '"' {
+			buffer = append(buffer, '"')
+			var result string
+			if err := json.Unmarshal(buffer, &result); err != nil {
+				return nil, p.makeError("invalid JSON string: %v", err)
 			}
+			n.raw = result
+			n.end = p.position
+			return n, nil
 		}
 
+		buffer = append(buffer, byte(c))
+	}
+}
+
+func (p *parser) tryUnicode(hex []rune) (string, bool) {
+	char, err := strconv.Unquote(fmt.Sprintf("'\\u%s'", string(hex)))
+	if err != nil {
+		return "", false
+	}
+	return char, true
+}
+
+func (p *parser) trySurrogate(hex []rune) (string, error) {
+	var surrogatePair []uint16
+
+	var hexPairs [][]rune
+	for _, hexPair := range append(hexPairs, hex[:4], hex[4:]) {
+		parsedPair, err := strconv.ParseUint(string(hexPair), 16, 16)
+		if err != nil {
+			return "", p.makeError("invalid unicode character '%s': %s", string(hexPair), err)
+		}
+		surrogatePair = append(surrogatePair, uint16(parsedPair))
+
+	}
+	decoded := utf16.Decode(surrogatePair)
+	return string(decoded), nil
+}
```

Comment:
We don't need them anymore.

Reply:
nice catch! removed in https://github.com/aquasecurity/jfather/pull/2/commits/8c15c46eb3f6644169316ba6a556285949214e87